### ISSUE TITLE
[Bug]: InMemoryPercentage is 100 after querycoord or standalone instance restart

### ISSUE THREAD

**Issue by congqixia** [milvus-io/milvus#17098]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version:b956817
- Deployment mode(standalone or cluster): both
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

If querycoord restarts, the collection in-memory percentage is still 100, which may be not in sync with the cluster true state.

### Expected Behavior

In memory percentage shall reflect the collection load status at all situation.

### Anything else?

Related to #17086

**Comment by congqixia** [comment]
Also related #17091 #17103

**Comment by congqixia** [comment]
See also #17048

**Comment by congqixia** [comment]
So the fixing plan is 

- [x] Make querynode load segment and loadChannels API idempotent #17109
- [ ] Make sure collection is loaded after load collection information from KV if the inMemory percentage is 100

**Comment by xiaofan-luan** [comment]
Make sure collection is loaded after load collection information from KV if the inMemory percentage is 100

what about querycoord just collect all information from alive querynode? think persistent progress info from KV is error prone

**Comment by congqixia** [comment]
> Make sure collection is loaded after load collection information from KV if the inMemory percentage is 100
> 
> what about querycoord just collect all information from alive querynode? think persistent progress info from KV is error prone

Yes. The `InMemoryPercentage` from KV cannot be trusted after the `Querycoord` restarts. Currently the plan is to make sure all load interfaces (segment/dml/delta channel) are idempotent and calling them asynchronously in previously stated situation. Before this procedure finishes, the collection load progress shall be 0 or unknown.

**Comment by letian-jiang** [comment]
The `InMemoryPercentage` cannot be trusted even if the querycoord keeps alive. Based on the current implementation, if one querynode crashed after collection loading, the search service is unavailable yet  `InMemoryPercentage` is still 100. (I have tested that it will recover after some time via load balance between querynodes).

A workaround is querycoord checks "if the collection is perfectly loaded" upon restarting. 
1. Set `InMemoryPercentage` to 0.
2. If perfectly loaded, we set the `InMemoryPercentage` to 100. 

But this workaround does not completely solve this `InMemoryPercentage` unreliable problem and brings extra complexities. 

To fully resolve this problem, there should be a joint design for memory replica, collection loading, and load balance. And the memory replica should not be used just for improving search throughput, but also to provide robustness against querynode crashes.

I think this issue needs further discussion. If it blocks any tests, I suggest that we rewrite these test cases to avoid this issue for now.

**Comment by congqixia** [comment]
> The `InMemoryPercentage` cannot be trusted even if the querycoord keeps alive. Based on the current implementation, if one querynode crashed after collection loading, the search service is unavailable yet `InMemoryPercentage` is still 100. (I have tested that it will recover after some time via load balance between querynodes).
> 
> A workaround is querycoord checks "if the collection is perfectly loaded" upon restarting.
> 
>     1. Set `InMemoryPercentage` to 0.
> 
>     2. If perfectly loaded, we set the `InMemoryPercentage` to 100.
> 
> 
> But this workaround does not completely solve this `InMemoryPercentage` unreliable problem and brings extra complexities.
> 
> To fully resolve this problem, there should be a joint design for memory replica, collection loading, and load balance. And the memory replica should not be used just for improving search throughput, but also to provide robustness against querynode crashes.
> 
> I think this issue needs further discussion. If it blocks any tests, I suggest that we rewrite these test cases to avoid this issue for now.

@letian-jiang I agree with most of your opinions. Just one thing: the behavior of the load progress is still 100% after cluster restarts need to be corrected.

**Comment by letian-jiang** [comment]
After some brainstorming with @congqixia, I summarize the main points as follows: 
* `InMemoryPercentage` should merely denote the progress of the very first collection loading. `InMemoryPercentage = 100` means the collection loading has been done. But it does not guarantee if the whole collection is currently loaded in memory. The system, however, may guarantee the "majority" of the collection is in-memory now. 
* Once we have loaded the collection, the Search/Query service should be available by default. If some in-memory segments/channels are lost due to querynode crash, we should automatically reload these segments/channels from object storage to memory. 
* If we enable the memory replica which replicates segments/channels in memory, the query service can be more available because the system provides some sort of redundancy. In this way, a querynode may also reload segments/channels from another querynode.

**Comment by letian-jiang** [comment]
A quick fix to catch up with the release of 2.1 would be
* proxy keeps retry search/query service until context timeout
* if timeout, return timeout exception indicating "the service is unavailable temporally, please retry later"
